src/nnet3/nnet-am-decodable-simple.h

```diff
@@ -392,8 +395,247 @@ class DecodableAmNnetSimpleParallel: public DecodableInterface {
+    for (BatchInfoMap::const_iterator iter =
+        batch_info_.begin(); iter != batch_info_.end(); iter++) {
+      if ((iter->second)->size() == minibatch_size_ )
+        return true;
+    }
+    return false;
+  }
+
+  // If true, it means the class has no data need to be computed. It is called
+  // from Compute() with "flush==true"
+  inline bool Empty() const {
+    for (BatchInfoMap::const_iterator iter =
+        batch_info_.begin(); iter != batch_info_.end(); iter++) {
+      if ((iter->second)->size() != 0)
+        return false;
+    }
+    return true;
+  }
+
+  // When an utterance is taken by GetFinishedUtterance(), clear its information
+  // in this class and release the memory on the heap.
+  void Clear(std::string utt_id);
+
+  // According to the information in batch_info_, prepare the 'batch' data,
+  // ComputationRequest, compute and get the results out.
+  void DoNnetComputation();
+  // If ensure_exact_final_context is true, this function is used to deal with
+  // "shorter than chunk size" utterances. In this function, we have to build
+  // a new CompuationRequest.
+  void DoNnetComputationOnes();
+
+  // Gets the iVector that will be used for this chunk of frames, if we are
+  // using iVectors (else does nothing).  note: the num_output_frames is
+  // interpreted as the number of t value, which in the subsampled case is not
+  // the same as the number of subsampled frames (it would be larger by
+  // opts_.frame_subsampling_factor).
+  void GetCurrentIvector(std::string utt_id,
+                         int32 output_t_start,
+                         int32 num_output_frames,
+                         Vector<BaseFloat> *ivector);
+
+  // called from constructor
+  void CheckAndFixConfigs();
+
+  // called from AcceptInput()
+  void CheckInput(const Matrix<BaseFloat> *feats,
+                  const Vector<BaseFloat> *ivector = NULL,
+                  const Matrix<BaseFloat> *online_ivectors = NULL);
+
+  // called from AcceptInput() or Compute(). Prepare the batch_info_ which
+  // will be used to compute.
+  void PrepareBatchInfo();
+
+  // called from constructor. According to (tot_left_context,tot_right_context),
+  // which equals to the model left/right context plus the extra left/right
+  // context, we prepare the frequently-used ComputationRequest.
+  // The CompuationRequest will be delete in deconstructor.
+  // Otherwise, we will initialize the "batch_info_" map which is used to
+  // maintain each information entry in (tot_left_context, tot_right_context)
+  // batch. The "batch_info_" map will be delete in deconstructor.
+  void PrepareComputationRequest();
+
+  NnetSimpleComputationOptions opts_;
+  const Nnet &nnet_;
+  int32 nnet_left_context_;
+  int32 nnet_right_context_;
+  int32 output_dim_;
+  // the log priors (or the empty vector if the priors are not set in the model)
+  CuVector<BaseFloat> log_priors_;
+
+  std::unordered_map<std::string, const Matrix<BaseFloat> * > feats_;
```

Comment:
I believe the default hasher for std::string is very bad, it just uses the length so it would hash them all to one bucket.  You should use StringHasher.